src/test/scala/edu/ie3/simona/model/em/opt/OptimizedFlexStratSpec.scala

```diff
@@ -0,0 +1,929 @@
+          batRes(1).pVal should approximate(10)
+          batRes(1).energyVal should approximate(1.75)
+
+          // discharging 10 kWh plus 2.5 kWh losses
+          batRes(2).pVal should approximate(-10)
+          batRes(2).energyVal should approximate(-10.75)
+
+          // charging 2 kWh minus 0.4 kWh losses
+          batRes(3).pVal should approximate(2)
+          batRes(3).energyVal should approximate(-9.15)
+
+        } withClue buildDebugString(batVars)
+
+        model.release()
+      }
+
+      "balance out additional power exceeding energy storage capacity" in {
+
+        given model: MPModel = MPModel(SolverLib.oJSolver)
+
+        val constFlex = EnergyBoundariesFlexOptions(
+          AssetEnergyBoundaries(
+            Seq(-10, -10, 10, 10).toPowerMap
+          )
+        )
+
+        val flexOptions = Map(
+          loadUUID -> constFlex,
+          batUUID -> batFlex,
+        )
+
+        val assetVars =
+          OptimizedFlexStrat.addAssetConstraints(flexOptions, sampleTime, ticks)
+
+        assetVars.toSeq should have length 2
+        assetVars.foreach(_.results should have length 1)
+        assetVars.foreach(_.results.foreach(_ should have length 4))
+
+        val objectiveContainer = OptimizedFlexStrat.buildObjective(
+          assetVars,
+          zeroKW,
+          MinAbsPowerObjectiveFactory,
+        )
+
+        model.minimize(objectiveContainer.objective)
+        model.start(timeLimit = 10000)
+
+        model.getStatus shouldBe SolutionStatus.OPTIMAL
+
+        /*
+          EXPECTED RESULTS
+          Since excess power costs the same at all points in time and
+          at all magnitudes, there are many optimal solutions.
+          Thus, we only test for things that are true for every optimal
+          solution: We know when the battery should be definitely
+          full/empty and how much energy was charged/discharged.
+
+          The soft constraints are vital here. Without them,
+          optimization would overestimate the losses in the first half
+          in order to achieve total power closer to zero.
+         */
+
+        val batVars = assetVars
+          .find(_.assetUuid == batUUID)
+          .getOrElse(fail(s"No asset variables for battery ($batUUID) found."))
+
+        val batRes = batVars.results.headOption
+          .getOrElse(fail(s"Empty results for battery ($batUUID)."))
+
+        {
+          objectiveContainer.softConstraints.foreach { constraint =>
+            withClue(constraint.getWarningMessage) {
+              constraint.getError should be < constraintTolerance
+            }
+          }
+
+          // possibly charging
+          batRes(0).pVal should be >= 0d
+          batRes(0).energyVal should (be >= 0d and be <= 12d)
+
+          // possibly charging, now we should have reached 24 kWh
+          // (12 kWh above starting energy)
+          batRes(1).pVal should be >= 0d
+          batRes(1).energyVal should approximate(12d)
```

Comment:
approx 8?

Reply:
As I mention above, this is the result of the second step already